framework/include/cppmicroservices/detail/TrackedBundle.tpp

```diff
@@ -0,0 +1,136 @@
+/*=============================================================================
+
+  Library: CppMicroServices
+
+  Copyright (c) The CppMicroServices developers. See the COPYRIGHT
+  file at the top-level directory of this distribution and at
+  https://github.com/CppMicroServices/CppMicroServices/COPYRIGHT .
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+=============================================================================*/
+
+namespace cppmicroservices {
+
+namespace detail {
+
+template<class TTT>
+TrackedBundle<TTT>::TrackedBundle(BundleTracker<T>* bundleTracker,
+                                  BundleTrackerCustomizer<T>* customizer)
+  : Superclass(bundleTracker->d->context)
+  , bundleTracker(bundleTracker)
+  , customizer(customizer)
+  , latch{}
+{
+}
+
+template<class TTT>
+void TrackedBundle<TTT>::WaitOnCustomizersToFinish()
+{
+  latch.Wait();
+}
+
+template<class TTT>
+void TrackedBundle<TTT>::BundleChanged(const BundleEvent& event)
+{
+  // Call track or untrack based on state mask
+
+  // Ignore events that do not correspond with
+  // Bundle state changes
+
+  (void)latch.CountUp();
```

Comment:
I don't think this is the correct use of the latch. See Declarative Services CCActiveState.cpp for the correct usage. 
 Example code:
 * function1()
 * {
 *    if(latch.CountUp())
 *    {
 *       // do something critical here
 *       latch.CountDown();
 *    }
 * }
 * function2()
 * {
 *    latch.Wait(); // this blocks the thread until all threads that have exited function1
 *    // do something here.
 * }
 
CountUp is protecting some code. It can only continue if the return value is true.